pow/src/compute/mod.rs

```diff
@@ -89,6 +89,50 @@ fn need_new_vm<M: randomx::WithCacheMode>(
 	need_new_vm
 }
 
+fn do_new_vm<M: randomx::WithCacheMode>(
+	key_hash: &H256,
+	machine: &RefCell<Option<(H256, randomx::VM<M>)>>,
+	shared_caches: &Mutex<LruCache<H256, Arc<randomx::Cache<M>>>>,
+) -> () {
+	let mut shared_caches = shared_caches.lock().expect("Poisoned mutex.");
+
+	if let Some(cache) = shared_caches.get_mut(key_hash) {
+		machine.replace(Some((*key_hash, randomx::VM::new(cache.clone(), global_config()))));
+	} else {
+		info!(
+			target: "kulupu-randomx",
+			"At block boundary, generating new RandomX {} cache with key hash {} ...",
+			M::description(),
+			key_hash,
+		);
+
+		let cache = match randomx::Cache::new(global_config()) {
+			Ok(mut cache) => {
+				cache.init(&key_hash[..]);
+				Arc::new(cache)
+			},
+			Err(_) => {
```

Comment:
Have you really checked this works in the situation of OOM? I think many OS will just kill the program instead of returning any error.

I suggest we change this to logic something like this:

* Detect if largepages is enabled.
* If not, use the current logic -- create a new cache.
* If so, we ensure there is always only **one single dataset** in memory. We do this by always reinit the current available cache, and never create any new pointers.

Reply:
> Have you really checked this works in the situation of OOM? I think many OS will just kill the program instead of returning any error.

Yes, i tried it with a limited amount of huge pages and was working as intended. Generally speaking the OS's heap allocation system call will just answer yes/no to an allocation request. Linux will start to kill procesesses with some heuristic above a threeshold of low free memory but i wouldn't say it's related to the answer of the allocation request, i mean it isn't synchronous... I could even check the behaviour in the case of large pages disabled, using a VM with a limited amount of ram! Usually in C and C++ it's an usual and recommended practise to always check the return of the allocation functions 🤔 

> I suggest we change this to logic something like this:
> 
>     * Detect if largepages is enabled.

👌 
>     * If not, use the current logic -- create a new cache.

👌 
>     * If so, we ensure there is always only **one single dataset** in memory. We do this by always reinit the current available cache, and never create any new pointers.

I'll think how i could do it in a clean a simple way 👌